keps/sig-node/4381-dra-semantic-parameters/README.md

```diff
@@ -0,0 +1,1477 @@
+    minimumMemory: 8Gi
+    sharing:
+      strategy: TimeSliced
+
+nodeName: worker-1
+partitionable:
+  instances:
+  - partitionable-gpu-2/half-and-quarters/second-half/quarter-0
+  - partitionable-gpu-2/half-and-quarters/second-half/quarter-1
+```
+
+## Design Details
+
+### ResourceClass extension
+
+A new, optional field in ResourceClass enables semantic parameters for claims
+using this class:
+
+```go
+type ResourceClass struct {
+    ...
+
+    // If (and only if) allocation of claims using this class is handled
+    // via semantic parameters, then SemanticParameters must be set to true.
+    SemanticParameters bool
+}
+```
+
+### NodeResourceSlice
+
+For each node, one or more NodeResourceSlice objects get created. The kubelet
+publishes them with the node as the owner, so they get deleted when a node goes
+down and then gets removed.
+
+All list types are atomic because that makes tracking the owner for
+server-side-apply (SSA) simpler. Patching individual list elements is not
+needed and there is a single owner (kubelet).
+
+```go
+// NodeResourceSlice provides information about available
+// resources on individual nodes.
+type NodeResourceSlice struct {
+    metav1.TypeMeta
+    // Standard object metadata
+    // +optional
+    metav1.ObjectMeta
+
+    Spec NodeResourceSliceSpec
+}
+```
+
+There is only a spec for a NodeResourceSlice at the moment. It holds the
+information about available resources. A status is not strictly needed because
+the information in the allocated claim statuses is sufficient to determine
+which of those resources are reserved for claims.
+
+However, despite the finalizer on the claims it could happen that a
+sufficiently determined user deletes a claim while it is in use. Therefore
+adding a status is a useful future extension. That status will include
+information about reserved resources (set by schedulers before allocating a
+claim) and in-use resources (set by kubelet). This then enables conflict
+resolution when multiple schedulers schedule pods to the same node because they
+would be required to set a reservation before proceeding with the
+allocation. It also enables detecting inconsistencies and taking actions to fix
+those, like deleting pods which use a deleted claim.
+
+```go
+type NodeResourceSliceSpec {
+    // NodeName identifies the node where the capacity is available.
+    // A field selector can be used to list only NodeResources
+    // objects with a certain node name.
+    NodeName string
+
+    // DriverName identifies the DRA driver providing the capacity information.
+    // A field selector can be used to list only NodeResources
+    // objects with a certain driver name.
+    DriverName string
+
+    NodeResourceModel // inline, field names must not conflict with the ones above
+}
+
+// NodeResourceModel must have one and only one field set.
+type NodeResourceModel struct {
+    Partitionable *PartitionableResources
+}
+
+type PartitionableResources struct {
```

Comment:
Repeating (sorry if it's tedious) - I am not sure on the details of this model, but it's OK for 1.30 to get the machinery all in place and test out the concept.

Reply:
I think we need some (somewhat realistic) model in the KEP, such as this one -- otherwise it's hard to see how all the pieces fit together. Maybe it could just be made more clear up front that the model used throughout the KEP is more of a motivating example of how one could model resources with this "semantic parameters" framework (but the finer details of the model's design / implementation are out of scope).

Reply:
I am kind of with Tim here. I worry that folks (like Tim did) will get really bogged down in the details of the model. Is there something simpler we can do in 1.30 to avoid that? I also worry that the *implementation* will get bogged down in making a sophisticated model work.

Reply:
> Maybe it could just be made more clear up front that the model used throughout the KEP is more of a motivating example of how one could model resources with this "semantic parameters" framework (but the finer details of the model's design / implementation are out of scope).

I agree on both points - but maybe then we can use a simpler model for that, just to not confuse people or get them going down a side path?

Reply:
I want mostly to avoid "anchoring" and constraining future-us to evolutions of past-us's decisions.  Future us should be ready and willing to discard everything past us did (and in fact should PLAN TO).

v1alpha2 can look nothing like v1alpha1, because it's alpha.

Reply:
>  I worry that folks (like Tim did) will get really bogged down in the details of the model. 

That is exactly why I had originally proposed three KEPs, with the intend that people reading the "semantic parameters" KEP should ignore details about the model and just treat the examples as, well, fictional examples. Specific models then would have been discussed in those separate KEPs, perhaps without including them in 1.30.

That didn't work ("how does this work for GPUs?", "here's how I would do it", with lots of back-and-forth) and led us to defining a model that is sufficient for GPUs directly in this KEP (the current state).

> we can use a simpler model for that

That was my goal with the "counter" model. The problem with that one was that it was too simplistic and wouldn't have been usable for anything real. I like @klueska [proposal from today](https://github.com/kubernetes/enhancements/pull/4384#issuecomment-1933616215) better: "device plugin-style list of resources with attributes attached to each" seems useful *and* simple enough that it's easy to understand and explain in this KEP.

Reply:
Here's a proposal for how to proceed: for each alternative semantic model, let's create one PR against [my branch]( https://github.com/pohly/enhancements/tree/dra-numeric-parameters) which modifies the README.md with that model.

@klueska: perhaps you can do that for "device plugin-style list of resources with attributes attached to each"?

Then we have some alternatives to choose from today and finalizing this KEP PR will be fast (merge+squash).

Reply:
Will do. I’ll do it with and without attributes (which would remove the debate about how to represent a selector).

Reply:
https://github.com/pohly/enhancements/pull/17/ is the version with attributes.

Reply:
I never ended up doing one without them.